File: lib/yarnlock.js
"use strict";

/**
 * Parser and serializer for yarn.lock, both dialects.
 *
 *   yarn v1 ("classic")        yarn v2+ ("berry")
 *   ------------------------   ---------------------------
 *   "pkg@^1.0.0":              "pkg@npm:^1.0.0":
 *     version "1.0.0"            version: 1.0.0
 *     resolved "https://..."     resolution: "pkg@npm:1.0.0"
 *     dependencies:              dependencies:
 *       other "^2.0.0"             other: "npm:^2.0.0"
 *
 * Both are an indentation-structured map from a package specifier to its
 * resolution -- the identity-keyed shape the merge engine is built for. The
 * separator differs: v1 puts a space between a scalar key and its value,
 * berry puts a colon. Berry keys contain colons of their own, so naive
 * splitting corrupts them.
 *
 * ROUND-TRIP FIDELITY IS THE HARD REQUIREMENT. If serializing an unchanged
 * document does not reproduce the original bytes, every merge produces a diff
 * touching the whole file, which is worse than the conflict it replaced.
 */

const V1 = "v1";
const BERRY = "berry";

class YarnLockError extends Error {}

/** Identify which yarn wrote this file, by markers yarn itself emits. */
function detectDialect(text) {
  if (text.includes("__metadata:")) return BERRY;
  if (text.includes("yarn lockfile v1")) return V1;
  // Guessing here is exactly the silent-corruption risk this module exists to
  // avoid: a wrong guess rewrites every separator in the file.
  throw new YarnLockError(
    "cannot identify yarn.lock dialect: no '__metadata:' or 'yarn lockfile v1' marker"
  );
}

/** Returns { header, doc, dialect }. `header` is the leading comment banner,
 *  preserved verbatim -- yarn writes a "manual changes might be lost" notice
 *  that must survive untouched. */
function parse(text) {
  const dialect = detectDialect(text);
  const lines = text.split("\n");

  const header = [];
  let i = 0;
  while (i < lines.length && (lines[i].trim() === "" || lines[i].trimStart().startsWith("#"))) {
    header.push(lines[i]);
    i++;
  }

  const { block } = parseBlock(lines, i, 0, dialect);
  return { header, doc: block, dialect };
}

function parseBlock(lines, start, indent, dialect) {
  const out = {};
  let i = start;
  while (i < lines.length) {
    const raw = lines[i];
    if (raw.trim() === "") {
      i++;
      continue;
    }
    const cur = raw.length - raw.replace(/^ +/, "").length;
    if (cur < indent) break;
    if (cur > indent) {
      // A deeper line with no parent header is malformed. Refusing beats
      // silently reparenting somebody's dependency.
      throw new YarnLockError(`unexpected indent at line ${i + 1}: ${JSON.stringify(raw)}`);
    }

    const stripped = raw.trim();
    if (stripped.endsWith(":") && !isScalar(stripped, dialect)) {
      const key = stripped.slice(0, -1);
      const res = parseBlock(lines, i + 1, indent + 2, dialect);
      out[key] = res.block;
      i = res.next;
    } else {
      const [key, value] = splitScalar(stripped, dialect, i);
      out[key] = value;
      i++;
    }
  }
  return { block: out, next: i };
}

/** Distinguish `dependencies:` (a block header) from `foo: bar` (a scalar). */
function isScalar(stripped, dialect) {
  if (dialect === V1) return false;
  return stripUotedPrefix(stripped).includes(": ");
}

function stripUotedPrefix(s) {
  if (!s.startsWith('"')) return s;
  const end = closingQuote(s, 0);
  return end === -1 ? s : s.slice(end + 1);
}

function splitScalar(stripped, dialect, lineno) {
  const sep = dialect === BERRY ? ": " : " ";
  const parts = partitionOutsideQuotes(stripped, sep);
  if (!parts) {
    throw new YarnLockError(
      `malformed ${dialect} entry at line ${lineno + 1}: ${JSON.stringify(stripped)}`
    );
  }
  return parts;
}

/** Split on the first `sep` that is not inside a quoted key. */
function partitionOutsideQuotes(s, sep) {
  if (s.startsWith('"')) {
    const end = closingQuote(s, 0);
    if (end !== -1) {
      const rest = s.slice(end + 1);
      if (rest.startsWith(sep)) return [s.slice(0, end + 1), rest.slice(sep.length)];
      return null;
    }
  }
  const idx = s.indexOf(sep);
  if (idx === -1) return null;
  return [s.slice(0, idx), s.slice(idx + sep.length)];
}

function closingQuote(s, start) {
  let i = start + 1;
  while (i < s.length) {
    if (s[i] === "\\") {
      i += 2;
      continue;
    }
    if (s[i] === '"') return i;
    i++;
  }
  return -1;
}

function serialize(header, doc, dialect) {
  const lines = header.slice();
  renderBlock(doc, 0, dialect, true, lines);
  return lines.join("\n");
}

function renderBlock(block, indent, dialect, top, lines) {
  const pad = " ".repeat(indent);
  let wrote = false;
  for (const key of Object.keys(block)) {
    const value = block[key];
    if (value !== null && typeof value === "object" && !Array.isArray(value)) {
      if (top && wrote) lines.push("");
      lines.push(`${pad}${key}:`);
      renderBlock(value, indent + 2, dialect, false, lines);
    } else {
      const sep = dialect === BERRY ? ": " : " ";
      lines.push(`${pad}${key}${sep}${value}`);
    }
    wrote = true;
  }
  if (top) lines.push("");
}

module.exports = { parse, serialize, detectDialect, V1, BERRY, YarnLockError };

File: src/yarnlock.py
"""Parser and serializer for yarn.lock, both dialects.

yarn.lock is the highest-value derived file not covered by a JSON parser, and
it comes in two incompatible shapes:

  yarn v1 ("classic")        yarn v2+ ("berry")
  ------------------------   ---------------------------
  "pkg@^1.0.0":              "pkg@npm:^1.0.0":
    version "1.0.0"            version: 1.0.0
    resolved "https://..."     resolution: "pkg@npm:1.0.0"
    dependencies:              dependencies:
      other "^2.0.0"             other: "npm:^2.0.0"

Both are an indentation-structured map from a package specifier to its
resolution -- exactly the identity-keyed shape the merge engine is built for.
The separator differs: v1 puts a space between a scalar key and its value,
berry puts a colon.

ROUND-TRIP FIDELITY IS THE HARD REQUIREMENT. If serializing an unchanged
document does not reproduce the original bytes, every merge produces a diff
touching the whole file, which is worse than the conflict it replaced. The
tests assert byte-identical round-trips on real 40,000-line lockfiles.
"""

from __future__ import annotations

V1 = "v1"
BERRY = "berry"


class YarnLockError(ValueError):
    pass


def detect_dialect(text: str) -> str:
    """Identify which yarn wrote this file.

    Checked against markers yarn itself emits rather than guessed from the
    body, because a wrong guess silently changes every separator on write.
    """
    if "__metadata:" in text:
        return BERRY
    if "yarn lockfile v1" in text:
        return V1
    # No marker. Berry always writes __metadata, so an unmarked file with
    # colon-separated scalars is still most likely berry-shaped; but guessing
    # here is exactly the silent-corruption risk this module exists to avoid.
    raise YarnLockError(
        "cannot identify yarn.lock dialect: no '__metadata:' or "
        "'yarn lockfile v1' marker"
    )


def parse(text: str):
    """Return (header_lines, document, dialect).

    `header_lines` are the leading comments, preserved verbatim -- yarn writes
    a "manual changes might be lost" banner that must survive untouched.
    """
    dialect = detect_dialect(text)
    lines = text.split("\n")

    header = []
    i = 0
    while i < len(lines) and (not lines[i].strip() or lines[i].lstrip().startswith("#")):
        header.append(lines[i])
        i += 1

    doc, consumed = _parse_block(lines, i, 0, dialect)
    return header, doc, dialect


def _parse_block(lines, start, indent, dialect):
    """Parse every entry at `indent`, returning (dict, next_line_index)."""
    out = {}
    i = start
    while i < len(lines):
        raw = lines[i]
        if not raw.strip():
            i += 1
            continue
        cur = len(raw) - len(raw.lstrip(" "))
        if cur < indent:
            break
        if cur > indent:
            # A deeper line with no parent header is malformed; refusing is
            # better than silently reparenting somebody's dependency.
            raise YarnLockError(f"unexpected indent at line {i + 1}: {raw!r}")

        stripped = raw.strip()
        if stripped.endswith(":") and not _is_scalar(stripped, dialect):
            key = stripped[:-1]
            child, i = _parse_block(lines, i + 1, indent + 2, dialect)
            out[key] = child
        else:
            key, value = _split_scalar(stripped, dialect, i)
            out[key] = value
            i += 1
    return out, i


def _is_scalar(stripped: str, dialect: str) -> bool:
    """Distinguish `dependencies:` (a block) from `foo: bar` (a scalar).

    A line ending in ':' is a block header only when nothing follows the
    colon. In berry a quoted key may itself contain a colon, so we check the
    unquoted tail rather than the raw string.
    """
    if dialect == V1:
        return False
    body = _strip_quoted_prefix(stripped)
    return ": " in body


def _split_scalar(stripped, dialect, lineno):
    if dialect == BERRY:
        key, sep, value = _partition_outside_quotes(stripped, ": ")
        if not sep:
            raise YarnLockError(f"malformed berry entry at line {lineno + 1}: {stripped!r}")
        return key, value
    key, sep, value = _partition_outside_quotes(stripped, " ")
    if not sep:
        raise YarnLockError(f"malformed v1 entry at line {lineno + 1}: {stripped!r}")
    return key, value


def _strip_quoted_prefix(s: str) -> str:
    """Return `s` with a leading quoted token removed, if there is one."""
    if not s.startswith('"'):
        return s
    end = _closing_quote(s, 0)
    return s[end + 1:] if end != -1 else s


def _partition_outside_quotes(s: str, sep: str):
    """Split on the first `sep` that is not inside a quoted key."""
    if s.startswith('"'):
        end = _closing_quote(s, 0)
        if end != -1:
            rest = s[end + 1:]
            if rest.startswith(sep):
                return s[: end + 1], sep, rest[len(sep):]
            return s, "", ""
    idx = s.find(sep)
    if idx == -1:
        return s, "", ""
    return s[:idx], sep, s[idx + len(sep):]


def _closing_quote(s: str, start: int) -> int:
    """Index of the quote closing the one at `start`, honouring backslashes."""
    i = start + 1
    while i < len(s):
        if s[i] == "\\":
            i += 2
            continue
        if s[i] == '"':
            return i
        i += 1
    return -1


def serialize(header, doc, dialect) -> str:
    out = list(header)
    body = _render_block(doc, 0, dialect, top=True)
    out.extend(body)
    return "\n".join(out)


def _render_block(block, indent, dialect, top=False):
    lines = []
    pad = " " * indent
    for key, value in block.items():
        if isinstance(value, dict):
            if top and lines:
                lines.append("")
            lines.append(f"{pad}{key}:")
            lines.extend(_render_block(value, indent + 2, dialect))
        else:
            sep = ": " if dialect == BERRY else " "
            lines.append(f"{pad}{key}{sep}{value}")
    if top:
        lines.append("")
    return lines
